File: store/111.二叉树的最小深度.js
/*
 * @lc app=leetcode.cn id=111 lang=javascript
 *
 * [111] 二叉树的最小深度
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
let minD;
const getMinDepth = (node) => {
  if (!node) return 0;
  if (node && !node.left && node.right) return getMinDepth(node.right) + 1;
  if (node && node.left && !node.right) return getMinDepth(node.left) + 1;
  return Math.min(getMinDepth(node.left), getMinDepth(node.right)) + 1;
};
let res;
const getPreMinDepth = (node, depth) => {
  if (!node) return;
  if (!node.left && !node.right) {
    res = res < depth ? res : depth;
    return;
  }
  if (node.left) getPreMinDepth(node.left, depth + 1);
  if (node.right) getPreMinDepth(node.right, depth + 1);
};
var minDepth = function (root) {
  // 法一
  // let queue = [],
  //   front,
  //   rear,
  //   minHeight;
  // front = rear = minHeight = 0;
  // if (!root) return minHeight;
  // queue[rear++] = root;
  // minHeight++;
  // while (front !== rear) {
  //   let size = rear - front;
  //   for (let i = 0; i < size; i++) {
  //     let node = queue[front++];
  //     if (node && !node.left && !node.right) return minHeight;
  //     if (node && node.left) queue[rear++] = node.left;
  //     if (node && node.right) queue[rear++] = node.right;
  //   }
  //   minHeight++;
  // }
  // 法二
  // return getMinDepth(root);
  // 法三
  if (!root) return 0;
  res = 9999;
  getPreMinDepth(root, 1);
  return res;
};
// @lc code=end

File: store/111.二叉树的最小深度.c
/*
 * @lc app=leetcode.cn id=111 lang=c
 *
 * [111] 二叉树的最小深度
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int result;
int back(struct TreeNode *root)
{
  // 前序
  // if (root->left == NULL && root->right == NULL)
  // {
  //   result = result > depth ? depth : result;
  //   return;
  // }

  // if (root->left != NULL)
  // {
  //   depth++;
  //   back(root->left, depth);
  //   depth--;
  // }
  // if (root->right != NULL)
  // {
  //   depth++;
  //   back(root->right, depth);
  //   depth--;
  // }
  if (root == NULL)
    return 0;
  int leftHeight = back(root->left);
  int rightHeight = back(root->right);
  if (root->left != NULL && root->right == NULL)
    return leftHeight + 1;
  if (root->right != NULL && root->left == NULL)
    return rightHeight + 1;
  return 1 + (leftHeight > rightHeight ? rightHeight : leftHeight);
}

int minDepth(struct TreeNode *root)
{
  // 法一：迭代
  // struct TreeNode **queue, *temp;
  // int front, rear, size, i, height;
  // queue = (struct TreeNode **)malloc(sizeof(struct TreeNode *) * 10000);
  // front = rear = height = 0;

  // if (root != NULL)
  // {
  //   queue[rear++] = root;
  //   while (front != rear)
  //   {
  //     size = rear;
  //     height++;
  //     for (i = front; i < size; i++)
  //     {
  //       temp = queue[front++];
  //       if (temp->left != NULL)
  //         queue[rear++] = temp->left;
  //       if (temp->right != NULL)
  //         queue[rear++] = temp->right;
  //       if (temp->left == NULL && temp->right == NULL)
  //         return;
  //     }
  //   }
  // }

  // return height;
  // 法二：递归
  // result = INT_MAX;
  // if (root != NULL)
  //   back(root, 1);
  // else
  //   result = 0;
  // return result;
  return back(root);
}
// @lc code=end
